From 89eeb4747448240166baf80dffa70f49c9eea00d Mon Sep 17 00:00:00 2001
From: Sergei Miroshnichenko <s.miroshnichenko@yadro.com>
Date: Wed, 13 May 2020 13:10:30 +0300
Subject: [PATCH 51/57] PCI: hotplug: movable bus numbers: Ignore bus numbers
 from bootloader

If a bridge was powered on and set up before being hot-added, ignore its
old bus numbers.

Signed-off-by: Sergei Miroshnichenko <s.miroshnichenko@yadro.com>
---
 drivers/pci/probe.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/pci/probe.c b/drivers/pci/probe.c
index 111111111..111111111 100644
--- a/drivers/pci/probe.c
+++ b/drivers/pci/probe.c
@@ -2921,6 +2921,14 @@ struct pci_dev *pci_scan_single_device(struct pci_bus *bus, int devfn)
 
 	pci_device_add(dev, bus);
 
+	if (pci_can_move_buses && dev->hdr_type == PCI_HEADER_TYPE_BRIDGE) {
+		u32 buses;
+
+		pci_read_config_dword(dev, PCI_PRIMARY_BUS, &buses);
+		buses &= 0xff000000;
+		pci_write_config_dword(dev, PCI_PRIMARY_BUS, buses);
+	}
+
 	return dev;
 }
 EXPORT_SYMBOL(pci_scan_single_device);

